hydro_lang/src/ir.rs

```diff
@@ -2239,6 +2263,105 @@ impl HydroNode {
+                    input.emit_core(builders_or_callback, built_tees, next_stmt_id);
+
+                let input_type = *input.metadata().output_type.clone().unwrap().0;
+                let (key_type, value_type) = if let syn::Type::Tuple(tuple) = input_type {
+                    assert_eq!(
+                        tuple.elems.len(),
+                        2,
+                        "ReduceKeyedWatermark input must be a tuple of (key, value)."
+                    );
+                    (tuple.elems[0].clone(), tuple.elems[1].clone())
+                } else {
+                    panic!("ReduceKeyedWatermark input must be a tuple of (key, value).");
+                };
+
+                let (watermark_ident, watermark_location_id) =
+                    watermark.emit_core(builders_or_callback, built_tees, next_stmt_id);
+
+                let cross_singleton_ident = syn::Ident::new(
+                    &format!("reduce_keyed_watermark_cross_singleton_{}", *next_stmt_id),
+                    Span::call_site(),
+                );
+
+                let fold_ident =
+                    syn::Ident::new(&format!("stream_{}", *next_stmt_id), Span::call_site());
+
+                match builders_or_callback {
+                    BuildersOrCallback::Builders(graph_builders) => {
+                        assert_eq!(
+                            input_location_id, watermark_location_id,
+                            "ReduceKeyedWatermark inputs must be in the same location"
+                        );
+
+                        let builder = graph_builders.entry(input_location_id).or_default();
+                        // 1. Don't allow any values to be added to the map if the key <= the watermark
+                        // 2. If the entry didn't exist in the BTreeMap, add it. Otherwise, call f.
+                        //    If the watermark changed, delete all BTreeMap entries with a key <= the watermark.
+                        // 3. Convert the BTreeMap back into a stream of (k, v)
+                        builder.add_dfir(
+                            parse_quote! {
+                                #cross_singleton_ident = cross_singleton();
+                                #input_ident -> [input]#cross_singleton_ident;
+                                #watermark_ident -> [single]#cross_singleton_ident;
+
+                                #fold_ident = #cross_singleton_ident
+                                    -> filter_map(|((k, v), watermark)| {
+                                        if k > watermark {
+                                            Some((k, v, watermark))
+                                        } else {
+                                            None
+                                        }
+                                    })
+                                    -> fold::<#lifetime>(|| (::std::collections::BTreeMap::<#key_type, #value_type>::new(), #key_type::default()), {
```

Comment:
This will not clear out old entries if just a watermark is sent but no new inputs arrive. In the original implementation, we `chain()`-ed the watermarks with inputs before sending them into the fold, instead of a `cross_singleton()`. We should be able to do that here as well.